### PR TITLE
Add tab for inspecting the state of Providers 

### PR DESCRIPTION
**Pull request by rrousselGit**

### CODE REVIEW

**Comment by jacob314** [comment]
Can you add a screenshots of what this looks like in action? That can make it easier to review as we know what to expect.

File: packages/devtools_app/lib/src/app.dart [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:devtools_app/src/provider/provider_screen.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/app.dart:5]: nit: Use the IDE quickfix to change this to a relative import

File: packages/devtools_app/lib/src/app.dart [hunk 2]
```diff
@@ -566,6 +567,10 @@ List<DevToolsScreen> get defaultScreens => <DevToolsScreen>[
         const VMDeveloperToolsScreen(),
         createController: () => VMDeveloperToolsController(),
       ),
+      DevToolsScreen<void>(
+        const ProviderScreen(),
+        createController: () {},
```
Comment (jacob314) [inline packages/devtools_app/lib/src/app.dart:572]: The irony... the provider screen doesn't have a provider for a controller yet. :)

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 3]
```diff
@@ -156,25 +158,52 @@ class EvalOnDartLibrary {
...
 
   Future<Instance> getInstance(
     FutureOr<InstanceRef> instanceRefFuture,
-    ObjectGroup isAlive,
+    Disposable isAlive,
   ) async {
     return await getObjHelper(await instanceRefFuture, isAlive);
   }
 
+  Future<Instance> getInstanceById(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:184]: These methods should be layered on top of EvalOnDartLibrary rather than embedded directly in EvalOnLibrary. EvalOnLibrary should be for evaluating in the context of any particular library not specific libraries such as widget_inspector or provider_binding.

File: packages/devtools_app/lib/src/inspector/inspector_service.dart [hunk 4]
```diff
@@ -556,12 +556,18 @@ class InspectorService extends DisposableController
   }
 }
 
+abstract class Disposable {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_service.dart:559]: Maybe move this out of InspectorService into eval_on_dart_library

File: packages/devtools_app/lib/src/provider/eval.dart [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+  } else {
+    return Result.error(
+      UnsupportedError('unkown type ${value.runtimeType}'),
+      StackTrace.current,
+    );
+  }
+}
+
+extension SafeEval on EvalOnDartLibrary {
+  Future<Instance> evalInstance(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/eval.dart:50]: Feel free to move this directly into EvalOnDartLibrary. Seems generally useful.

File: packages/devtools_app/lib/src/provider/eval.dart [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+  /// ```dart
+  /// Future<T> $await<T>(Future<T> future, String id) async {
+  ///   try {
+  ///     return await future;
+  ///   } finally {
+  ///     postEvent('future_completed', {'id': id});
+  ///   }
+  /// }
+  /// ```
+  Future<InstanceRef> awaitEval(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/eval.dart:82]: Same for this nice helper. Can you also provide a link to the dartsdk bug you filed about providing a way to wait for futures to complete as part of the vm service api?

File: packages/devtools_app/lib/src/provider/eval.dart [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+  static int _nextId = 0;
+
+  /// a [safeEval] variant that does not complete until the Future evaluated completes
+  ///
+  /// The expression **must** return a [Future].
+  ///
+  /// Works only if the evaluated library includes the following code:
+  ///
+  /// ```dart
+  /// Future<T> $await<T>(Future<T> future, String id) async {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/eval.dart:74]: btw you can make this a little more general purpose by injecting $await as part of the expression you are evaluating. that way you can use it in any library.
For example you can evaluate

"""
(() {
  Future<T> _await<T>(....) {...}
  _await($yourRealExpressionHere);
})()
"""
to inject _await dynamically. Just strip line breaks for it to work.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 8]
```diff
@@ -0,0 +1,805 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:1]: 2020 -> 2021

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 9]
```diff
@@ -0,0 +1,805 @@
...
+import '../eval_on_dart_library.dart';
+import '../globals.dart';
+import '../inspector/inspector_text_styles.dart';
+import '../inspector/inspector_tree.dart';
+import '../screen.dart';
+import '../split.dart';
+import 'eval.dart';
+import 'provider_state_controller.dart';
+
+const kTypeColor = Color.fromARGB(255, 78, 201, 176);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:23]: don't prefix constant names with k

Interested to see how these colors look in light and dark mode.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 10]
```diff
@@ -0,0 +1,805 @@
...
+const kBoolColor = Color.fromARGB(255, 86, 156, 214);
+const kNullColor = kBoolColor;
+const kNumColor = Color.fromARGB(255, 181, 206, 168);
+const kStringColor = Color.fromARGB(255, 206, 145, 120);
+const kPropertyColor = Color.fromARGB(255, 206, 145, 120);
+
+const double rowHeight = 20.0;
+const double horizontalSpacing = 15;
+const double expandIconSize = 20;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:32]: would be nice to keep the icons a consistent size with the other icons in devtools. there should be an existing const for this.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 11]
```diff
@@ -0,0 +1,805 @@
...
+    object: (instance) => instance.evalForInstance,
+    orElse: () => ref.watch(evalProvider),
+  );
+});
+
+class ProviderScreen extends Screen {
+  const ProviderScreen()
+      : super.conditional(
+          id: id,
+          requiresLibrary: 'package:provider/',
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:87]: nit: if there is a specific library only available under package:provider when you provide the support this screen needs you could make it conditional on that library.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 12]
```diff
@@ -0,0 +1,805 @@
...
+    return Split(
+      axis: splitAxis,
+      initialFractions: const [0.33, 0.67],
+      children: [
+        const _SplitBorder(child: _ProviderList()),
+        Column(
+          children: [
+            const Expanded(child: _SplitBorder(child: _ProviderValue())),
+            if (watch(_hasSelectedProviderProvider)) ...[
+              const SizedBox(height: 10),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:116]: Nit: use one of the default sizes in /theme.dart
That way the app will stay a bit more consistent and if we tweak the sizes all ui will be updated.
I imagine you want either `denseSpacing` or `defaultSpacing`
```
const defaultButtonHeight = 36.0;
const buttonMinWidth = 36.0;

const defaultIconSize = 16.0;
const actionsIconSize = 20.0;
const defaultIconThemeSize = 24.0;

const defaultSpacing = 16.0;
const denseSpacing = 8.0;
const denseRowSpacing = 6.0;

const defaultBorderRadius = 4.0;

const borderPadding = 2.0;
const densePadding = 4.0;

const smallProgressSize = 12.0;

const defaultListItemHeight = 28.0;

const defaultChartHeight = 150.0;

const defaultScrollBarOffset = 10.0;
```

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 13]
```diff
@@ -0,0 +1,805 @@
...
+      await serviceManager.performHotReload();
+    } catch (err) {
+      _showErrorSnackBar(context, err);
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.all(8.0),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:184]: use padding constant.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 14]
```diff
@@ -0,0 +1,805 @@
...
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          const Text('Execute code against the value exposed by this provider'),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:188]: I've been playing around with a Console window exposed across most screens in DevTools. Similarly to the console in chrome devtools, the last N objects selected would be available as $0, $1, ... $6 in the console. Fine to land this separately for now but I'd be interested in unifying so there is one console experience across the app.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 15]
```diff
@@ -0,0 +1,805 @@
...
+    final state = watch(providerIdsProvider);
+
+    return state.when(
+      loading: () => const Center(child: CircularProgressIndicator()),
+      error: (err, stack) => Center(
+        child: Text('Error $err\n\n$stack'),
+      ),
+      data: (providerNodes) {
+        return Scrollbar(
+          child: ListView.builder(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:220]: nit: use always visible scrollbars to make it feel more like a desktop app.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 16]
```diff
@@ -0,0 +1,805 @@
...
+      error: _buildError,
+      data: (instance) sync* {
+        final isExpanded = watch(_isExpandedProvider(path));
+        yield _buildHeader(
+          instance,
+          isExpanded: isExpanded,
+          disableExpand: disableExpand,
+        );
+
+        if (isExpanded.state) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:358]: this seems like it reimplementing the value display for the Debugger view. Lets see if we can integrate this functionality into the general debugger view instead or layer it on top of the debugger view which already deals with list, map, set, etc.

Comment (jacob314) [reply packages/devtools_app/lib/src/provider/provider_screen.dart:358]: The existing value display in the debugger view is fairly basic at the moment. I'd be thrilled to see it include some of the functionality it looks like you have added here.

File: packages/devtools_app/lib/src/provider/provider_screen.dart [hunk 17]
```diff
@@ -0,0 +1,805 @@
...
+      // A provider will automatically be selected as soon as one is detected
+      return Container();
+    }
+
+    return Scrollbar(
+      child: ListView.custom(
+        // TODO: item height should be based on font size
+        itemExtent: rowHeight,
+        padding: const EdgeInsets.symmetric(
+          vertical: 10,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/provider_screen.dart:593]: use devtools theme constants for padding

File: packages/devtools_app/lib/src/provider/result.dart [hunk 18]
```diff
@@ -0,0 +1,49 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'result.freezed.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/provider/result.dart:3]: Can you add a comment on how to run the code generator for this?

File: packages/devtools_app/pubspec.yaml [hunk 19]
```diff
@@ -57,9 +60,11 @@ dependencies:
 
 dev_dependencies:
   build_runner: ^1.3.0
+  freezed: ^0.12.0
```
Comment (jacob314) [inline packages/devtools_app/pubspec.yaml:63]: Fyi @kenzieschmoll this would add a dependency on package freezed.

**Review by jacob314 — COMMENTED** [review]
Looks very promising! Added some comments for discussion.

**Comment by rrousselGit** [comment]
> Can you add a screenshots of what this looks like in action? That can make it easier to review as we know what to expect.

Sure! 

image



Although I'm working on some golden tests, so that should makes things easier

File: packages/devtools_app/lib/src/provider/provider_screen.dart (hunk 11 above)
Comment (rrousselGit) [reply packages/devtools_app/lib/src/provider/provider_screen.dart:87]: Only provider is required, but it needs a specific version constraints (probably >=5.0.0-nullsafety.x)

File: packages/devtools_app/lib/src/provider/provider_screen.dart (hunk 16 above)
Comment (rrousselGit) [reply packages/devtools_app/lib/src/provider/provider_screen.dart:358]: You mean the `Variables` panel inside the Debugger tab?

**Comment by jacob314** [comment]
I'd like to discuss how we can show this content in the inspector instead of its own tab.  It could either be an extra tab next to the existing Layout Explorer and Details Tree tabs or we could get more clever. I think this should be the default content shown when a Provider widget is selected in the inspector.

**Comment by rrousselGit** [comment]
I'm not sure about that.

I'd prefer for this view to be fairly independent from the rest, both code and UI wise, to simplify the maintenance – especially as I will have to maintain two tabs at once (one for [provider](http://pub.dev/packages/provider) and one for [riverpod](http://pub.dev/packages/riverpod))

What I'd really like would be to ultimately implement a plugin-system for tabs (similar to build_runner), so that the code of this tab could be in a separate repository (ideally versioned with provider).

**Comment by jacob314** [comment]
> I'm not sure about that.
> 
> I'd prefer for this view to be fairly independent from the rest, both code and UI wise, to simplify the maintenance – especially as I will have to maintain two tabs at once (one for [provider](http://pub.dev/packages/provider) and one for [riverpod](http://pub.dev/packages/riverpod))
> 
> What I'd really like would be to ultimately implement a plugin-system for tabs (similar to build_runner), so that the code of this tab could be in a separate repository (ideally versioned with provider).
Discussed offline and keeping in its own tab is fine. I would like to also help reuse the shared console view instead of having a custom console view for this tab. Use case seems the same as needing the shared console view to work well for the inspector. I would like to see the object debugger viewer from this screen basically become the default object debugger view for all debugger views. It isn't really provider specific and we should have one view for debugging and editing properties of objects rather than multiple different views that basically solve the same task.
Agreed long term it would be nice to have a plugin system for tabs.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 20]
```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
...
+    Disposable isAlive,
   ) async {
     return await getObjHelper(await instanceRefFuture, isAlive);
   }
 
+  Future<String> getInstanceHashCode(
+    InstanceRef instance, {
+    @required Disposable isAlive,
+  }) async {
+    // TODO(rrousselGit) figure out why sometimes this returns null
```
Comment (rrousselGit) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:190]: I have seen cases wherea `getInstance` returns null even when the return is non-null (and there's no exception).
I haven't been able to reliably reproduce it, so I'm not sure what causes it.

File: packages/devtools_app/lib/main.dart [hunk 21]
```diff
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/main.dart:6]: how much value are we getting from riverpod here? Using both provider and riverpod seems a bit redundant.

File: packages/devtools_app/lib/src/banner_messages.dart [hunk 22]
```diff
@@ -254,6 +254,50 @@ Relaunch your application with the '--profile' argument, or ''',
...
+  BannerMessage build(BuildContext context) {
+    return _BannerError(
+      key: Key('ProviderUnknownErrorBanner - $screenId'),
+      screenId: screenId,
+      textSpans: const [
+        TextSpan(
+          text: '''
+The devtool failed to connect with package:provider.
+
+This could be caused by an outdated version of package:provider. Make sure that you are using a version >=5.0.0.''',
```
Comment (jacob314) [inline packages/devtools_app/lib/src/banner_messages.dart:271]: Instead of guessing, can you tweak package:provider so you can tell exactly which version it is from code?
For example, you could check for the presence of an API not present before 5.0.0 using EvalOnLibrary or store a string in the library that lists the version.

File: packages/devtools_app/lib/src/banner_messages.dart [hunk 23]
```diff
@@ -254,6 +254,50 @@ Relaunch your application with the '--profile' argument, or ''',
   }
 }
 
+class ProviderUnknownErrorBanner {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/banner_messages.dart:257]: both these messages are specific to your screens so put them in the directory of the screen rather than the generic messages file.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 24]
```diff
@@ -56,14 +63,16 @@ class EvalOnDartLibrary {
...
-  Completer<LibraryRef> _libraryRef;
   Future<void> _initializeComplete;
   StreamSubscription selectedIsolateStreamSubscription;
 
   String get isolateId => _isolateId;
   String _isolateId;
 
+  Completer<LibraryRef> _libraryRef;
   Future<LibraryRef> get libraryRef => _libraryRef.future;
+  Completer<LibraryRef> get libraryRefCompleter => _libraryRef;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:74]: do you really need the completer rather than just the Future? Completer should be an implementation detail.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 20 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:190]: has your objectGroup already been disposed when that occurs?

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 25]
```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
...
   ) async {
     return await getObjHelper(await instanceRefFuture, isAlive);
   }
 
+  Future<String> getInstanceHashCode(
+    InstanceRef instance, {
+    @required Disposable isAlive,
+  }) async {
+    // TODO(rrousselGit) figure out why sometimes this returns null
+    final hash = await evalInstance(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:191]: If you had the identity hash would that be enough?
@bkonyi would be really handy if all InstanceRef objects had the identity hash code.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 26]
```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
...
+  /// ```dart
+  /// Future<T> $await<T>(Future<T> future, String id) async {
+  ///   try {
+  ///     return await future;
+  ///   } finally {
+  ///     postEvent('future_completed', {'id': id});
+  ///   }
+  /// }
+  /// ```
+  Future<InstanceRef> awaitEval(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:229]: Fyi @bkonyi it would be handy if this was supported by the VM Service so we didn't have to hack it.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 27]
```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
...
+      '''
+() async { try { return await $expression; } finally { postEvent('future_completed', {'id': $id}); } }()
+''',
+      isAlive: isAlive,
+      scope: scope,
+    );
+
+    await future;
+
+    return result;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:253]: This gives you the Future that has completed but how do you get the actual value it completed with?

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 28]
```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
...
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    try {
+      final result = await addRequest(
+        isAlive,
+        () => _safeEval(expression, scope: scope),
+      );
+      if (isAlive.disposed) {
+        throw StateError(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:268]: why is this an error? shouldn't it just mean that the request has been cancelled and we should return a null InstanceRef?

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 29]
```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
...
+    Map<String, String> scope,
+  }) async {
+    if (disposed) {
+      throw StateError(
+        'Called `safeEval` on a disposed `EvalOnDartLibrary` instance',
+      );
+    }
+
+    // copied from `_eval`
+    String libraryRefId;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:291]: please reuse instead of duplicating this.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 25 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:191]: [Still waiting for a review on that change](https://dart-review.googlesource.com/c/sdk/+/188600).

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 30]
```diff
@@ -272,3 +436,83 @@ class LibraryNotFound implements Exception {
...
+
+  final ErrorRef errorRef;
+  final String expression;
+  final Map<String, String> scope;
+
+  @override
+  String toString() {
+    return 'Evaluation `$expression` failed with $errorRef';
+  }
+}
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:518]: I would suggest keeping the changes to devtools core functionality in this CL and moving changes that are adding new pages to a separate CL. that will make the review easier.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 26 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:229]: Can you file a feature request and assign to me?

File: packages/devtools_app/lib/src/instance_viewer/eval.dart [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+
+import '../eval_on_dart_library.dart';
+import '../globals.dart';
+
+/// An [EvalOnDartLibrary] that has access to no specific library in particular
+///
+/// Not suitable to be used when evaluating third-party objects, as it would
+/// otherwise not be possible to read private properties.
+final evalProvider = Provider<EvalOnDartLibrary>((ref) {
+  final eval = EvalOnDartLibrary(['dart:io'], serviceManager.service);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/instance_viewer/eval.dart:18]: why is this 'dart:io'? perhaps use 'dart:core'?

File: packages/devtools_app/lib/src/result.dart [hunk 32]
```diff
@@ -0,0 +1,50 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+// run `flutter pub run build_runner build --delete-conflicting-outputs` to re-generate
+part 'result.freezed.dart';
+
+@freezed
+abstract class Result<T> with _$Result<T> {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/result.dart:7]: Document what this class is and why it is useful?
If it is only used for the Provider screen and riverpod screens move it under that directory.

File: packages/devtools_app/lib/src/riverpod/riverpod_screen.dart [hunk 33]
```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/riverpod/riverpod_screen.dart:1]: nit: wrong year.

File: packages/devtools_app/lib/src/sliver_iterable_child_delegate.dart [hunk 34]
```diff
@@ -0,0 +1,39 @@
+import 'package:flutter/widgets.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/sliver_iterable_child_delegate.dart:1]: Fyi @kenzieschmoll @terrylucas. This could be handy other cases in DevTools as well.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart [hunk 35]
```diff
@@ -171,7 +171,7 @@ class VmServiceWrapper implements VmService {
   Future get onDone => _vmService.onDone;
 
   @override
-  Future<void> dispose() => _vmService.dispose();
+  Future<void> dispose() async => _vmService.dispose();
```
Comment (jacob314) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:174]: Was this change accidental?

File: packages/devtools_app/macos/Flutter/GeneratedPluginRegistrant.swift [hunk 36]
```diff
@@ -2,6 +2,8 @@
 //  Generated file. Do not edit.
 //
 
+// clang-format off
```
Comment (jacob314) [inline packages/devtools_app/macos/Flutter/GeneratedPluginRegistrant.swift:5]: revert or move this to a separate CL.

File: packages/devtools_app/pubspec.yaml [hunk 37]
```diff
@@ -11,18 +11,17 @@ version: 0.9.8-dev.1
 homepage: https://github.com/flutter/devtools
 
 environment:
-  sdk: '>=2.10.0 <3.0.0'
+  sdk: ">=2.10.0 <3.0.0"
```
Comment (jacob314) [inline packages/devtools_app/pubspec.yaml:14]: switch back to single quotes

File: packages/devtools_app/pubspec.yaml [hunk 38]
```diff
@@ -62,12 +63,12 @@ dev_dependencies:
   devtools_testing: 0.9.8-dev.1
   flutter_test:
     sdk: flutter
+  freezed: ^0.12.0
```
Comment (jacob314) [inline packages/devtools_app/pubspec.yaml:66]: Fyi @kenzieschmoll for evaluation how much of an issue these extra package imports would cause in g3.

File: packages/devtools_app/test/memory_screen_test.dart [hunk 39]
```diff
@@ -58,6 +57,7 @@ void main() {
     expect(find.byType(MemoryBody), findsOneWidget);
   }
 
+  // Set a wide enough screen width that we do not run into overflow.
```
Comment (jacob314) [inline packages/devtools_app/test/memory_screen_test.dart:60]: move unrelated cleanup like this to a separate CL so this main one is easier to review. Code reviews are generally O(n^2) in the number of lines and files to review so if we can break this into a few <2000 line CLs instead of one 10,000 line CL that will help.

File: packages/devtools_testing/fixtures/provider_app/.gitignore [hunk 40]
```diff
@@ -0,0 +1,46 @@
+# Miscellaneous
```
Comment (jacob314) [inline packages/devtools_testing/fixtures/provider_app/.gitignore:1]: move all these fixtures to a separate CL and I'll give them a quick LGTM. That will also help move the main CL down to a reviewable size.

File: packages/devtools_testing/lib/riverpod/controller_test.dart [hunk 41]
```diff
@@ -0,0 +1,190 @@
...
+      @required String provider,
+      Map<String, String> scope,
+    }) async {
+      final subscriptionRef = await eval.safeEval(
+        'container.listen($provider)',
+        isAlive: isAlive,
+        scope: {...scope, 'container': containerRef.id},
+      );
+
+      await eval.awaitEval('tester.pump()', isAlive: isAlive);
```
Comment (jacob314) [inline packages/devtools_testing/lib/riverpod/controller_test.dart:50]: add a comment explaining why you need to call tester.pump on the app being tested.

**Review by jacob314 — COMMENTED** [review]
Completed a first pass. Lets move the fixtures and the implementation of the provider and riverpod specific CLs over into separate CLs.
So far I've reviewed everything but the provider and riverpod page impls.
In general, I'll be pickier on stuff in the shared devtools components and more flexible on stuff just under the provider and riverpod screen directories.

**Comment by rrousselGit** [comment]
> CLs

What does this means?

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 25 above)
Comment (rrousselGit) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:191]: I am rendering what Flutter shows when using Diagnosticable (hash in hex)

**Comment by kenzieschmoll** [comment]
> >CLs

> What does this means?
> edit: separate PRs I guess? Not sure what the letters means though

Yes, you are correct. CL = change list (this is the google equivalent of PR so sometimes we are guilty of using them interchangeably)

File: packages/devtools_app/pubspec.yaml [hunk 42]
```diff
@@ -123,12 +124,13 @@ flutter:
...
-  devtools_shared:  #OVERRIDE_FOR_DEVELOPMENT
-    path: ../devtools_shared  #OVERRIDE_FOR_DEVELOPMENT
-  devtools_testing:  #OVERRIDE_FOR_DEVELOPMENT
-    path: ../devtools_testing  #OVERRIDE_FOR_DEVELOPMENT
+    path: ../devtools_server #OVERRIDE_FOR_DEVELOPMENT
+  devtools_shared: #OVERRIDE_FOR_DEVELOPMENT
+    path: ../devtools_shared #OVERRIDE_FOR_DEVELOPMENT
+  devtools_testing: #OVERRIDE_FOR_DEVELOPMENT
+    path: ../devtools_testing #OVERRIDE_FOR_DEVELOPMENT
+  freezed_annotation: ^0.12.0
```
Comment (kenzieschmoll) [inline packages/devtools_app/pubspec.yaml:136]: Can this be removed?

**Comment by rrousselGit** [comment]
Closing in favor of https://github.com/flutter/devtools/pull/2851, which is a branch that contains only the `package:provider` tab
